File: 173_Binary_Search_Tree_Iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:
    Node = None 
    listOfValues = None 

    def __init__(self, root: TreeNode):
        self.Node = root 
        self.listOfValues = self.traverseTree(root, [])
    
    def traverseTree(self, root, listOfValues):
        if(root == None):
            return listOfValues 
        #in-order (left - root -right) 
        if(root.left != None):
            listOfValues = self.traverseTree(root.left, listOfValues)
        listOfValues.append(root.val)
        if(root.right != None):
            listOfValues = self.traverseTree(root.right, listOfValues)
        return listOfValues
        

    def next(self) -> int:
        return self.listOfValues.pop(0)
        

    def hasNext(self) -> bool:
        return len(self.listOfValues) > 0
        


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: 173_Binary_Search_Tree_Iterator.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {
public:
    vector<int> listOfNodes; 
    
    BSTIterator(TreeNode* root) {
        listOfNodes = traverse(root, {}); 
    }
    
    vector<int> traverse(TreeNode* root, vector<int> result)
    {
        if(root == NULL) return result; 
        if(root->left != NULL) result = traverse(root->left, result);
        result.push_back(root->val); 
        if(root->right != NULL) result = traverse(root->right, result); 
        return result; 
    }
    
    int next() {
        int front = listOfNodes.front();
        listOfNodes.erase(listOfNodes.begin());
        return front; 
    }
    
    bool hasNext() {
        return listOfNodes.size() > 0; 
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */